File: sources/academy/webscraping/scraping_basics_javascript/exercises/guardian_f1_authors.mjs
import * as cheerio from 'cheerio';

async function download(url) {
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error(`HTTP ${response.status}`);
  }
  const html = await response.text();
  return cheerio.load(html);
}

const listingUrl = 'https://www.theguardian.com/sport/formulaone';
const $ = await download(listingUrl);

const promises = $('#maincontent ul li').toArray().map(async (element) => {
  const $item = $(element);
  const $link = $item.find('a').first();
  if (!$link.length) {
    return;
  }

  const articleUrl = new URL($link.attr('href'), listingUrl).href;
  const $article = await download(articleUrl);

  const title = $article('h1').text().trim();
  if (!title) {
    return;
  }

  const author = $article('a[rel="author"]').first().text().trim();
  const attribution = author || $article('aside address').first().text().trim() || 'null';

  console.log(`${attribution}: ${title}`);
});

await Promise.all(promises);

File: sources/academy/webscraping/scraping_basics_python/exercises/guardian_f1_authors.py
import httpx
from bs4 import BeautifulSoup
from urllib.parse import urljoin


def download(url: str) -> BeautifulSoup:
    response = httpx.get(url)
    response.raise_for_status()
    return BeautifulSoup(response.text, "html.parser")


def parse_author(article_soup: BeautifulSoup) -> str | None:
    link = article_soup.select_one('a[rel="author"]')
    if link:
        return link.text.strip()
    address = article_soup.select_one('aside address')
    if address:
        return address.text.strip()
    return None


listing_url = "https://www.theguardian.com/sport/formulaone"
listing_soup = download(listing_url)

for item in listing_soup.select('#maincontent ul li'):
    link = item.select_one('a')
    if not link or 'href' not in link.attrs:
        continue

    article_url = urljoin(listing_url, link['href'])
    article_soup = download(article_url)

    title = article_soup.select_one('h1').text.strip()
    author = parse_author(article_soup)

    print(f"{author}: {title}")
